datafusion-jit/src/api.rs

```diff
@@ -0,0 +1,630 @@
+        }
+    }
+
+    /// Add one more parameter to the function.
+    pub fn param(mut self, name: impl Into<String>, ty: JITType) -> Self {
+        let name = name.into();
+        assert!(!self.fields.back().unwrap().contains_key(&name));
+        self.params.push((name.clone(), ty));
+        self.fields.back_mut().unwrap().insert(name, ty);
+        self
+    }
+
+    /// Set return type for the function. Functions are of `void` type by default if
+    /// you do not set the return type.
+    pub fn ret(mut self, name: impl Into<String>, ty: JITType) -> Self {
+        let name = name.into();
+        assert!(!self.fields.back().unwrap().contains_key(&name));
+        self.ret = Some((name.clone(), ty));
+        self.fields.back_mut().unwrap().insert(name, ty);
+        self
+    }
+
+    /// Enter the function body at start the building.
+    pub fn enter_block(&mut self) -> CodeBlock {
+        self.fields.push_back(HashMap::new());
+        CodeBlock {
+            fields: &mut self.fields,
+            state: &self.assembler_state,
+            stmts: vec![],
+            while_state: None,
+            if_state: None,
+            fn_state: Some(GeneratedFunction {
+                name: self.name.clone(),
+                params: self.params.clone(),
+                body: vec![],
+                ret: self.ret.clone(),
+            }),
+        }
+    }
+}
+
+/// Keep `while` condition expr as we are constructing while loop body.
+struct WhileState {
+    condition: Expr,
+}
+
+/// Keep `if-then-else` state, including condition expr, the already built
+/// then statements (if we are during building the else block).
+struct IfElseState {
+    condition: Expr,
+    then_stmts: Vec<Stmt>,
+    in_then: bool,
+}
+
+impl IfElseState {
+    /// Move the all current statements in the `then` block and move to `else` block.
+    fn enter_else(&mut self, then_stmts: Vec<Stmt>) {
+        self.then_stmts = then_stmts;
+        self.in_then = false;
+    }
+}
+
+/// Code block that
+pub struct CodeBlock<'a> {
+    /// A stack that containing all defined variables so far. The variables defined
+    /// in the current block are at the top stack frame.
+    /// Fields provides a shadow semantics of the same name in outsider block, and are
+    /// used to guarantee type safety while constructing AST.
+    fields: &'a mut VecDeque<HashMap<String, JITType>>,
+    /// The state of Assembler, used for type checking function calls.
+    state: &'a Arc<Mutex<AssemblerState>>,
+    /// Holding all statements for the current code block.
+    stmts: Vec<Stmt>,
+    while_state: Option<WhileState>,
+    if_state: Option<IfElseState>,
+    /// Keep track of function params and return types, only valid for function main block.
+    fn_state: Option<GeneratedFunction>,
+}
+
+impl<'a> CodeBlock<'a> {
+    pub fn build(&mut self) -> GeneratedFunction {
+        assert!(
+            self.fn_state.is_some(),
+            "Calling build on a non function block"
```

Comment:
It may not clear what is "non function block". Maybe just say `fn_state` is empty?

Reply:
A code block can be the code block of the whole function, a block inside `while`, `if-then` clauses, or `else` clauses.  But it's only valid to call the `build` method on the `function` block since it's used to seal all codes for the generated function.

Reply:
And only the outermost `function` block has `fn_state` non_empty, which is used to keep functions parameters and returns.

Reply:
then we can just use "Can only call build on outermost function block" for the error message? "non function block" looks quite unclear.

Reply:
or "Calling build on a block that is not function block"?